### PR TITLE
Adding a DeepInv solver

### PR DESCRIPTION
**Pull request by wassimmazouz**
- Added a dataset which uses DeepInverse methods for noise.
- Implemented a grayscaling function for RGB torch images.
- Added a DeepInverse solver which implements the 2D TV Inpainting problem
- Adapted `objective.py` to use the ad-hoc linear operator `linop()` in case of a DeepInv-generated dataset.

### CODE REVIEW

File: datasets/chateau.py [hunk 1]
```diff
@@ -0,0 +1,39 @@
+from benchopt import BaseDataset, safe_import_context
+
+with safe_import_context() as import_ctx:
+    import deepinv as dinv
+    import torch
```
Comment (tomMoral) [inline datasets/chateau.py]: I would avoid using deepinv for the dataset, this makes it harder to run this dataset without torch, which is for most other solvers.
You can create a linear operator A without deepinv.

Also, performing the operation on GPU for such small scale computation (processing the dataset) is not a very good idea as the overhead of transferring the image `x` to GPU and back in memory is probably larger than the gain in time.

File: objective.py [hunk 2]
```diff
@@ -3,33 +3,49 @@
...
+        else:
+            device = 'cpu'
+        operator = dinv.physics.Inpainting(
+            tensor_size=size[1:],
+            mask=0.5,
+            device=device
+        )
+        out = operator(x).squeeze(0)
+        out = out.squeeze(0)
+        return out.numpy()
```
Comment (tomMoral) [inline objective.py:37]: This should come from the dataset. Note that you can create a LinearOperator that applies the transform as `A @ u` with [scipy `LinearOperator` API](https://docs.scipy.org/doc/scipy/reference/generated/scipy.sparse.linalg.LinearOperator.html#scipy.sparse.linalg.LinearOperator)

File: solvers/deepinv.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    def set_objective(self, A, reg, delta, data_fit, y, isotropy):
+        self.A, self.reg, self.y = A, reg, torch.from_numpy(y)
+        self.delta, self.data_fit = delta, data_fit
+        self.isotropy = isotropy
+
+    def run(self, n_iter):
+        if torch.cuda.is_available():
+            device = dinv.utils.get_freer_gpu()
+        else:
+            device = 'cpu'
```
Comment (tomMoral) [inline solvers/deepinv.py:32]: Here this make sense to use GPU.

File: solvers/deepinv.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+        elif isotropy == 'anisotropic':
+            return True, f"solver does not work with {isotropy} regularization"
+        return False, None
+
+    def set_objective(self, A, reg, delta, data_fit, y, isotropy):
+        self.A, self.reg, self.y = A, reg, torch.from_numpy(y)
+        self.delta, self.data_fit = delta, data_fit
+        self.isotropy = isotropy
+
+    def run(self, n_iter):
```
Comment (tomMoral) [inline solvers/deepinv.py:28]: Use the [callback strategy](https://benchopt.github.io/user_guide/performance_curves.html#using-a-callback) here

File: solvers/deepinv.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+            mask=0.5,
+            device=device
+        )
+        physics.noise_model = dinv.physics.GaussianNoise(sigma=0.2)
+        for _ in range(n_iter):
+            x = x - data_fidelity.grad(x, y, physics)
+            if self.recon == 'TV':
+                x = prior.prox(x,  gamma=reg)
+            elif self.recon == 'DRUNet':
+                x = denoiser(x, reg)
```
Comment (tomMoral) [inline solvers/deepinv.py:59]: Could you use a step size here?
I think for inpainting the lipschitz constant is 1  so `lr=1` should be working but it might be different for other problems.

File: solvers/deepinv.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        self.out = x.clone()
+        self.out = self.out.squeeze(0)
+        self.out = self.out.squeeze(0)
```
Comment (tomMoral) [inline solvers/deepinv.py:62]: why not just `squeeze` once?
```suggestion
        self.out = self.out.squeeze()
```

File: solvers/deepinv.py [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+            denoiser = dinv.models.DRUNet(
+                in_channels=1, out_channels=1, pretrained='download',
+                device=device)
+            prior = dinv.optim.PnP(denoiser=denoiser)
+
+        physics = dinv.physics.Inpainting(
+            tensor_size=x.shape[1:],
+            mask=0.5,
+            device=device
+        )
```
Comment (tomMoral) [inline solvers/deepinv.py:52]: When you generate a new `Inpainting` object, it will differ from the one you used to generate the data and for the objective.
This should be generated once in the dataset.

Note that if you generate a for image inpainting, you can directly [pass the mask when constructing the inpainting object](https://deepinv.github.io/deepinv/stubs/deepinv.physics.Inpainting.html#deepinv.physics.Inpainting).

**Review by tomMoral — COMMENTED** [review]
Hello @wassimmazouz 

Nice PR! A few comments on the way we handle data generation and on the dependencies.
Let me know if you have questions about it. :)
